GitUI/CommandsDialogs/BrowseDialog/FormBrowseMenus.cs

```diff
@@ -115,6 +132,122 @@ static ToolStripItem CreateItem(ToolStrip senderToolStrip)
+                        AppSettings.SetBool(toolbarPrefix + separator.Name, separator.Visible);
+                        shouldHideNextSeparator = true;
+                    }
+                    else
+                    {
+                        shouldHideNextSeparator &= !item.Visible;
+                    }
+                }
+
+                // Second pass: toolbar items from right to left
+                shouldHideNextSeparator = true;
+                for (int i = senderToolStrip.Items.Count - 1; i >= 0; i--)
+                {
+                    var item = senderToolStrip.Items[i];
+
+                    if (item is ToolStripSeparator separator)
+                    {
+                        if (!separator.Visible)
+                        {
+                            continue;
+                        }
+
+                        separator.Visible = !shouldHideNextSeparator;
+                        AppSettings.SetBool(toolbarPrefix + separator.Name, separator.Visible);
+                        shouldHideNextSeparator = true;
+                    }
+                    else
+                    {
+                        shouldHideNextSeparator &= !item.Visible;
+                    }
+                }
+            }
+
+            string? currentGroup = null;
+            foreach (ToolStripItem toolbarItem in senderToolStrip.Items)
+            {
+                if (toolbarItem is ToolStripSeparator)
+                {
+                    toolStripItem.DropDownItems.Add(new ToolStripSeparator());
+                    continue;
+                }
+
+                string text = toolbarItem.ToolTipText;
+
+                if (BelongToAGroup(toolbarItem, out string groupname))
+                {
+                    if (currentGroup == groupname)
+                    {
+                        toolbarItem.Visible = AppSettings.GetBool(toolbarPrefix + toolbarItem.Name, true);
+                        continue;
+                    }
+
+                    currentGroup = (string)toolbarItem.Tag;
+                }
+
+                var menuToolbarItem = new ToolStripMenuItem(text) { Checked = toolbarItem.Visible, CheckOnClick = true, Tag = toolbarItem, Image = toolbarItem.Image };
+
+                toolbarItem.Visible = AppSettings.GetBool(toolbarPrefix + toolbarItem.Name, true);
+                menuToolbarItem.Click += (s, e) =>
+                {
+                    toolbarItem.Visible = !toolbarItem.Visible;
+                    AppSettings.SetBool(toolbarPrefix + toolbarItem.Name, toolbarItem.Visible);
+
+                    if (BelongToAGroup(toolbarItem, out _))
+                    {
+                        SetVisibilityForSameGroupToolbarItem(toolbarItem.Tag, toolbarItem.Visible);
+                    }
+
+                    AdaptSeparatorsVisibility();
+                };
+
+                toolStripItem.DropDownItems.Add(menuToolbarItem);
+            }
+
+            AdaptSeparatorsVisibility();
```

Comment:
Keep all local methods at the end of the containing method, and after `return;` statement.

Reply:
Ok. I already doing some more refactoring to please CodeFactor....